AppDB/appscale/datastore/fdb/utils.py

```diff
@@ -0,0 +1,281 @@
+import json
+import logging
+import random
+import time
+
+import fdb
+import mmh3
+import six
+from tornado import gen
+from tornado.concurrent import Future as TornadoFuture
+
+from appscale.datastore.dbconstants import InternalError, SCATTER_CHANCE
+from appscale.datastore.fdb.codecs import Path
+
+fdb.api_version(610)
+logger = logging.getLogger(__name__)
+
+MAX_FDB_TX_DURATION = 5
+
+_MAX_SEQUENTIAL_BIT = 52
+_MAX_SEQUENTIAL_ID = (1 << _MAX_SEQUENTIAL_BIT) - 1
+_MAX_SCATTERED_COUNTER = (1 << (_MAX_SEQUENTIAL_BIT - 1)) - 1
+_MAX_SCATTERED_ID = _MAX_SEQUENTIAL_ID + 1 + _MAX_SCATTERED_COUNTER
+_SCATTER_SHIFT = 64 - _MAX_SEQUENTIAL_BIT + 1
+
+# The Cloud Datastore API uses microseconds as version IDs. When the entity
+# doesn't exist, it reports the version as "1".
+ABSENT_VERSION = 1
+
+# The max number of bytes for each FDB value.
+CHUNK_SIZE = 10000
+
+SCATTER_PROP = u'__scatter__'
+
+MAX_32 = 256 ** 4
+
+SCATTER_PROPORTION = int(MAX_32 * SCATTER_CHANCE)
+
+# The number of bytes used to store a commit versionstamp.
+VERSIONSTAMP_SIZE = 10
+
+MAX_ENTITY_SIZE = 1048572
+
+# The FDB directory used for the datastore.
+DS_ROOT = (u'appscale', u'datastore')
+
+
+class FDBErrorCodes(object):
+  NOT_COMMITTED = 1020
+
+
+def ReverseBitsInt64(v):
+  v = ((v >> 1) & 0x5555555555555555) | ((v & 0x5555555555555555) << 1)
+  v = ((v >> 2) & 0x3333333333333333) | ((v & 0x3333333333333333) << 2)
+  v = ((v >> 4) & 0x0F0F0F0F0F0F0F0F) | ((v & 0x0F0F0F0F0F0F0F0F) << 4)
+  v = ((v >> 8) & 0x00FF00FF00FF00FF) | ((v & 0x00FF00FF00FF00FF) << 8)
+  v = ((v >> 16) & 0x0000FFFF0000FFFF) | ((v & 0x0000FFFF0000FFFF) << 16)
+  v = int((v >> 32) | (v << 32) & 0xFFFFFFFFFFFFFFFF)
+  return v
+
+
+class ScatteredAllocator(object):
+  def __init__(self):
+    self._counter = random.randint(1, _MAX_SCATTERED_COUNTER)
+
+  def invalidate(self):
+    self._counter = random.randint(1, _MAX_SCATTERED_COUNTER)
+
+  def get_id(self):
+    id_ = (_MAX_SEQUENTIAL_ID + 1 +
+           long(ReverseBitsInt64(self._counter << _SCATTER_SHIFT)))
+
+    self._counter += 1
+    if self._counter > _MAX_SCATTERED_COUNTER:
+      self._counter = 1
+
+    return id_
+
+
+class TornadoFDB(object):
+  def __init__(self, io_loop):
+    self._io_loop = io_loop
+
+  @gen.coroutine
+  def commit(self, tr, convert_exceptions=True):
+    tornado_future = TornadoFuture()
+    callback = lambda fdb_future: self._handle_fdb_result(
+      fdb_future, tornado_future)
+    commit_future = tr.commit()
+    commit_future.on_ready(callback)
+    try:
+      yield tornado_future
+    except fdb.FDBError as fdb_error:
+      if convert_exceptions:
+        raise InternalError(fdb_error.description)
```

Comment:
My understanding of InternalError is that it's used as a fallback when either no other error codes are suitable or there's some kind of unexpected failure. Since transactions that use commit versionstamps are not idempotent, I'm being a little conservative about which FDB errors we retry. At the moment, we only retry when there's an FDB concurrency error and only in certain cases. Later on, we can find more edge cases that would be suitable for a retry or a more specific error code.